core/src/main/java/org/apache/iceberg/util/TableScanUtil.java

```diff
@@ -78,4 +83,57 @@ public static CloseableIterable<CombinedScanTask> planTasks(CloseableIterable<Fi
             splitFiles),
         BaseCombinedScanTask::new);
   }
+
+  public static <T extends SplittableScanTask<T>> CloseableIterable<ScanTaskGroup<T>> planTaskGroups(
+      CloseableIterable<T> tasks,
+      long splitSize, int lookback, long openFileCost) {
+
+    Preconditions.checkArgument(splitSize > 0, "Invalid split size (negative or 0): %s", splitSize);
+    Preconditions.checkArgument(lookback > 0, "Invalid split planning lookback (negative or 0): %s", lookback);
+    Preconditions.checkArgument(openFileCost >= 0, "Invalid file open cost (negative): %s", openFileCost);
+
+    // capture manifests which can be closed after scan planning
+    CloseableIterable<T> splitTasks = CloseableIterable.combine(
+        FluentIterable.from(tasks).transformAndConcat(input -> input.split(splitSize)),
+        tasks);
+
+    Function<T, Long> weightFunc = task -> Math.max(task.totalSizeBytes(), task.totalFilesCount() * openFileCost);
+
+    return CloseableIterable.transform(
+        CloseableIterable.combine(
+            new BinPacking.PackingIterable<>(splitTasks, splitSize, lookback, weightFunc, true),
+            splitTasks),
+        combinedTasks -> new BaseScanTaskGroup<>(combineAdjacentTasks(combinedTasks)));
+  }
+
+  private static <T extends SplittableScanTask<T>> List<T> combineAdjacentTasks(List<T> tasks) {
+    if (tasks.isEmpty()) {
+      return tasks;
+    }
+
+    List<T> combinedTasks = Lists.newArrayList();
+    T lastTask = null;
+
+    for (T task : tasks) {
+      if (lastTask != null) {
+        if (lastTask.isAdjacent(task)) {
+          // merge with the last task
+          lastTask = lastTask.combineWithAdjacentTask(task);
```

Comment:
Maybe change `combineWithAdjacentTask(ThisT)` to `combineWithAdjacentTask(ScanTask)`. Then it is up to the implementation whether it can combine.

Reply:
Done.